### DISCUSSION TITLE
2021-3-8 Uva 514 Rails - 解题报告（欧伟凯）

### DISCUSSION THREAD

**Discussion by Owaiter** [CSGrandeur/s-1problem1day1ac#123] [Solutions]
## 题目大意

有n辆火车，行驶中途有个中转站，火车按1，2，3，······，n的顺序进站，在每一辆火车进站前，中转站停留的火车中最晚进站的火车可以选择离开中转站，其他火车均无法行动。现在问你，是否可以得到某个出站的序列，如果可以得到，输出Yes；不能得到，输出No。

## 解题思路

这道题很明显考的就是栈的使用，中转站便是一个栈，火车进站出站符合栈的先进后出规则，所以直接用栈模拟即可。

先用数组a[1005]存储目标序列，j表示从左到右序列判断的位置，然后将数字1到n依次进栈，每次有数字进栈，便用while循环判断a[j]是否等于栈顶的元素，如果相等，则j+1并将栈顶元素弹出，不等则继续将下一个数字进栈，当然，如果栈为空，也停止判断。当操作完n个数字之后，如果栈为空，说明可以实现该序列；反之，无法实现。

## 算法与复杂度分析

本题并没有使用到什么算法知识，单纯考栈的使用，如果对于栈还不是很了解的同学，可以[点击这里](https://blog.csdn.net/qq_42410605/article/details/100546720?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522161494501216780261956635%2522%252C%2522scm%2522%253A%252220140713.130102334..%2522%257D&request_id=161494501216780261956635&biz_id=0&utm_medium=distribute.pc_search_result.none-task-blog-2~all~sobaiduend~default-1-100546720.pc_search_result_before_js&utm_term=stack)，学习一下栈的知识。

由于需将n个数字依次进栈，并且每次都需判断栈顶元素是否匹配，所以最坏的情况就是前n-1个数进栈且判断时每次栈顶的元素都不匹配，此时计算的次数为2n，最后一个数n进栈时，栈里的n个元素依次弹出，此时计算的次数为3n，所以本题的时间复杂度为O(n)。

## 心得体会

本题的输入得仔细看，需要用到两层while循环，第一层输入序列数字的个数n，当n为0时退出循环，第二层需先输入序列的第一个数，并判断是否为0，若为0，则跳出第二层循环，一开始题没看完整直接看样例还是挺懵的。

## 代码

``` c++
#include<cstdio>
#include<stack>
#include<cstring>
int a[1005];
int main()
{
    int n;
    while(scanf("%d",&n) && n)
    {
        while(scanf("%d",&a[1]) && a[1])
        {
            std::stack<int>s;
            for(int i = 2;i <= n;i++)
                scanf("%d",&a[i]);
            int j = 1; //表示序列判断的位置
            for(int i = 1;i <= n;i++)
            {
                s.push(i);
                while(!s.empty() && a[j] == s.top()) { s.pop();j++; }
            }
            if(s.empty())
                printf("Yes\n");
            else
                printf("No\n");
        }
        printf("\n");
    }
    return 0;
}
```

**Comment by CSGrandeur** [comment]
给个不用stl的代码：
```cpp
#include<stdio.h>
#include<string.h>
#include<stdlib.h>
int stk[1100], stp;
int main()
{
    int n, num, ith;
    while(scanf("%d", &n) && n)
    {
        while(scanf("%d", &num) && num)
        {
            stp = ith = 0;
            for(int i = 0; i < n; i ++)
            {
                if(i) scanf("%d", &num);
                while(stp == 0 || stk[stp] < num)
                    stk[++ stp] = ++ ith;
                if(stk[stp] == num)
                    stp --;
            }
            printf(stp ? "No\n" : "Yes\n");
        }
        puts("");
    }
    return 0;
}
```
